src/tools/translate-to-fuzz.h

```diff
@@ -0,0 +1,1106 @@
+  Index labelIndex;
+
+  // a list of things relevant to computing the odds of an infinite loop,
+  // which we try to minimize the risk of
+  std::vector<Expression*> hangStack;
+
+  std::map<WasmType, std::vector<Index>> typeLocals; // type => list of locals with that type
+
+  void addFunction() {
+    Index num = wasm.functions.size();
+    func = new Function;
+    func->name = std::string("func_") + std::to_string(num);
+    func->result = getReachableType();
+    assert(typeLocals.empty());
+    Index numParams = upTo(upTo(5));
+    for (Index i = 0; i < numParams; i++) {
+      auto type = getConcreteType();
+      typeLocals[type].push_back(func->params.size());
+      func->params.push_back(type);
+    }
+    Index numVars = upTo(upTo(10));
+    for (Index i = 0; i < numVars; i++) {
+      auto type = getConcreteType();
+      typeLocals[type].push_back(func->params.size() + func->vars.size());
+      func->vars.push_back(type);
+    }
+    labelIndex = 0;
+    assert(breakableStack.empty());
+    assert(hangStack.empty());
+    // with reasonable chance make the body a block
+    if (oneIn(2)) {
+      func->body = makeBlock(func->result);
+    } else {
+      // with very small chance, make the body unreachable
+      if (oneIn(20)) {
+        func->body = make(unreachable);
+      } else {
+        func->body = make(func->result);
+      }
+    }
+    if (HANG_LIMIT > 0) {
+      func->body = builder.makeSequence(
+        makeHangLimitCheck(),
+        func->body
+      );
+    }
+    assert(breakableStack.empty());
+    assert(hangStack.empty());
+    wasm.addFunction(func);
+    // export some, but not all (to allow inlining etc.). make sure to
+    // export at least one, though, to keep each testcase interesting
+    if (num == 0 || oneIn(2)) {
+      auto* export_ = new Export;
+      export_->name = func->name;
+      export_->value = func->name;
+      export_->kind = ExternalKind::Function;
+      wasm.addExport(export_);
+    }
+    // cleanup
+    typeLocals.clear();
+  }
+
+  Name makeLabel() {
+    return std::string("label$") + std::to_string(labelIndex++);
+  }
+
+  // always call the toplevel make(type) command, not the internal specific ones
+
+  int nesting = 0;
+
+  Expression* make(WasmType type) {
+    // when we should stop, emit something small (but not necessarily trivial)
+    if (finishedInput ||
+        nesting >= 5 * NESTING_LIMIT || // hard limit
+        (nesting >= NESTING_LIMIT && !oneIn(3))) {
+      if (isConcreteWasmType(type)) {
+        if (oneIn(2)) {
+          return makeConst(type);
+        } else {
+          return makeGetLocal(type);
+        }
+      } else if (type == none) {
+        if (oneIn(2)) {
+          return makeNop(type);
+        } else {
+          return makeSetLocal(type);
+        }
+      }
+      assert(type == unreachable);
+      return makeTrivial(type);
+    }
+    nesting++;
+    Expression* ret;
+    switch (type) {
+      case i32: ret = __makei32(); break;
```

Comment:
Double-underscores are reserved for C++ implementations right? Is single-underscore sufficient here?